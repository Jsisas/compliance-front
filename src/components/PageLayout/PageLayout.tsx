import React, { useEffect, useState } from 'react';
import { Dropdown, Layout, Menu } from 'antd';
import { Link } from 'react-router-dom';
import { Routes } from '../../pages/Routes';
import { ReactComponent as Logo } from './../../assets/logo/small_logo.svg';
import styles from './PageLayout.module.scss';
import { useDispatch, useSelector } from 'react-redux';
import { RootState } from '../../redux/reducer';
import { clearAuth, getAuthentication } from '../../redux/Auth/AuthSlice';
import { authenticate } from '../../redux/Auth/AuthService';

const { Content, Sider } = Layout;

export function PageLayout(): JSX.Element {
	const homePage = '/regulations';
	const [selectedKey, setSelectedKey] = useState(getSelectedKey());
	const [isCollapsed, setCollapsed] = useState(false);
	const authentication = useSelector((state: RootState) => getAuthentication(state));
	const isAuthenticated = useSelector((state: RootState) => state.auth.isAuthenticated);

	const dispatch = useDispatch();
	useEffect(() => {
		if (!isAuthenticated && authentication && authentication.token) {
			dispatch(authenticate(authentication.token));
		}
	}, [dispatch, isAuthenticated, authentication]);

	useEffect(() => {
		backButtonListener();
	});

	function backButtonListener() {
		window.addEventListener('popstate', () => {
			setSelectedKey(getSelectedKey());
		});
	}

	function getSelectedKey() {
		return window.location.pathname === '/' ? homePage : window.location.pathname;
	}

	function handleLogOut() {
		dispatch(clearAuth());
	}

	const profileOptions = (
		<Menu>
			<Menu.Item key='Settings'>Settings</Menu.Item>
			<Menu.Item key='Account'>Account</Menu.Item>
			<Menu.Item key='Log out' onClick={() => handleLogOut()}>
				Log out
			</Menu.Item>
		</Menu>
	);

	return (
		<Layout>
			{isAuthenticated && (
				<Sider
					breakpoint='lg'
					collapsedWidth='0'
					theme={'light'}
					className={styles.extraExtraLightGrey}
					onCollapse={(collapsed) => setCollapsed(collapsed)}
					style={{ minHeight: '100vh' }}
				>
					<div className={styles.logo}>
						<Link to='/regulations' onClick={() => setSelectedKey('/regulations')}>
							<Logo />
						</Link>
					</div>
					<Menu className={styles.sideMenu} mode='inline' selectedKeys={[selectedKey]}>
						<Menu.Item key={'/regulations'} onClick={() => setSelectedKey('/regulations')}>
							<Link to='/regulations' className='nav-text'>
								Overview
							</Link>
						</Menu.Item>
						<Menu.Item key='/controls' onClick={() => setSelectedKey('/controls')}>
							<Link to='/controls' className='nav-text'>
								Controls
							</Link>
						</Menu.Item>
						<Menu.Item key='/tasks' onClick={() => setSelectedKey('/tasks')}>
							<Link to='/tasks' className='nav-text'>
								Tasks
							</Link>
						</Menu.Item>
					</Menu>
					<div className={styles.profile} style={{ display: isCollapsed ? 'none' : '', width: '175px' }}>
						<Dropdown overlay={profileOptions} trigger={['click']}>
							<img src={authentication.user.picture} alt='Avatar' />
						</Dropdown>
						<Link to={'/'}>{authentication.user.name}</Link>
					</div>
				</Sider>
			)}
			<Layout style={{ background: '#fff', minWidth: '340px' }} className={styles.white}>
				<Content className={styles.white}>
					<div className='site-layout-background' style={{ padding: 24, minHeight: 360 }}>
						<Routes />
					</div>
				</Content>
			</Layout>
		</Layout>
	);
}
