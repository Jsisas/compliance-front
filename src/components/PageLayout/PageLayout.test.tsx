import { deepRender } from '../../util/test/TestUtil';
import { screen } from '@testing-library/react';
import React from 'react';
import { testUsers } from '../../util/test/data/TestUsers';
import { PageLayout } from './PageLayout';

const initialAuthentication = { token: 'asd', user: testUsers[0] };

describe('<PageLayout />', () => {
	beforeEach(() => {
		localStorage.setItem('auth', JSON.stringify(initialAuthentication));
	});

	it('it matches a snapshot', () => {
		const [history, element] = deepRender(<PageLayout />, { auth: [initialAuthentication] });
		expect(element).toMatchSnapshot();
	});

	it('it can log out', () => {
		const [history] = deepRender(<PageLayout />, { auth: [initialAuthentication] });

		const profilePicture = screen.getByAltText('Avatar');
		profilePicture.click();

		const logoutButton = screen.getByText('Log out');
		logoutButton.click();
		expect(localStorage.getItem('auth')).toBeNull();
		expect(history.location.pathname).toBe('/login');
	});

	it('it highlights correct menu item on load', () => {
		deepRender(<PageLayout />, { auth: [initialAuthentication] });
		const overviewMenuItem = screen.getByText('Overview');
		expect(overviewMenuItem.parentElement).toHaveClass('ant-menu-item-selected');
	});

	it('it highlights correct menu item on click', () => {
		deepRender(<PageLayout />, { auth: [initialAuthentication] });
		const controlsMenuItem = screen.getByText('Controls');
		expect(controlsMenuItem.parentElement).not.toHaveClass('ant-menu-item-selected');
		controlsMenuItem.click();
		expect(controlsMenuItem.parentElement).toHaveClass('ant-menu-item-selected');
	});

	it('it hides side menu when on login page', () => {
		const [history, wrappedElement] = deepRender(<PageLayout />, undefined, '/login');
		expect(history.location.pathname).toBe('/login');

		expect(wrappedElement.queryByText('Overview')).not.toBeInTheDocument();
		expect(wrappedElement.getByText('Login')).toBeInTheDocument();
	});

	it('it hides side menu when user not logged in', () => {
		localStorage.clear();
		deepRender(<PageLayout />);
		expect(screen.queryByText('Overview')).not.toBeInTheDocument();
		expect(screen.getByText('Login')).toBeInTheDocument();
	});
});
