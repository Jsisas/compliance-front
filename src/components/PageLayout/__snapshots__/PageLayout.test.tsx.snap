// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PageLayout /> it matches a snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <section
        class="ant-layout ant-layout-has-sider"
        data-testid="page-layout"
      >
        <aside
          class="extraExtraLightGrey ant-layout-sider ant-layout-sider-light"
          style="min-height: 100vh; flex: 0 0 200px; max-width: 200px; min-width: 200px; width: 200px;"
        >
          <div
            class="ant-layout-sider-children"
          >
            <div
              class="logo"
            >
              <a
                href="/regulations"
              >
                <svg>
                  small_logo.svg
                </svg>
              </a>
            </div>
            <ul
              class="ant-menu sideMenu ant-menu-light ant-menu-root ant-menu-inline"
              role="menu"
            >
              <li
                class="ant-menu-item ant-menu-item-only-child ant-menu-item-selected"
                role="menuitem"
                style="padding-left: 24px;"
              >
                <a
                  class="nav-text"
                  href="/regulations"
                >
                  Overview
                </a>
              </li>
              <li
                class="ant-menu-item ant-menu-item-only-child"
                role="menuitem"
                style="padding-left: 24px;"
              >
                <a
                  class="nav-text"
                  href="/controls"
                >
                  Controls
                </a>
              </li>
              <li
                class="ant-menu-item ant-menu-item-only-child"
                role="menuitem"
                style="padding-left: 24px;"
              >
                <a
                  class="nav-text"
                  href="/tasks"
                >
                  Tasks
                </a>
              </li>
            </ul>
            <div
              class="profile"
              style="width: 175px;"
            >
              <img
                alt="Avatar"
                class="ant-dropdown-trigger"
              />
              <a
                href="/"
              >
                Liis Kindel
              </a>
            </div>
          </div>
        </aside>
        <section
          class="ant-layout white"
          style="background: rgb(255, 255, 255); min-width: 340px;"
        >
          <main
            class="ant-layout-content white"
          >
            <div
              class="site-layout-background"
              style="padding: 24px; min-height: 360px;"
            >
              <div
                class="ant-row ant-row-space-between"
                style="margin: -8px -8px 8px -8px;"
              >
                <div
                  class="ant-col ant-col-xs-24 ant-col-sm-16 ant-col-md-10 ant-col-lg-16 ant-col-xl-16 ant-col-xxl-16"
                  style="padding: 8px 8px 8px 8px;"
                >
                  <h1
                    class="ant-typography"
                  >
                    Overview
                  </h1>
                </div>
              </div>
              <div
                class="ant-row ant-row-space-between"
                style="margin: -8px -8px 8px -8px;"
              >
                <div
                  class="ant-col ant-col-xs-24 ant-col-sm-24 ant-col-md-24 ant-col-lg-24 ant-col-xl-24 ant-col-xxl-24"
                  style="padding: 8px 8px 8px 8px;"
                >
                  <div
                    class="ant-table-wrapper"
                    style="width: 100%;"
                  >
                    <div
                      class="ant-spin-nested-loading"
                    >
                      <div
                        class="ant-spin-container"
                      >
                        <div
                          class="ant-table ant-table-empty"
                        >
                          <div
                            class="ant-table-container"
                          >
                            <div
                              class="ant-table-content"
                            >
                              <table
                                style="table-layout: auto;"
                              >
                                <colgroup />
                                <thead
                                  class="ant-table-thead"
                                >
                                  <tr>
                                    <th
                                      class="ant-table-cell"
                                    />
                                  </tr>
                                </thead>
                                <tbody
                                  class="ant-table-tbody"
                                >
                                  <tr
                                    class="ant-table-placeholder"
                                  >
                                    <td
                                      class="ant-table-cell"
                                    >
                                      <div
                                        class="ant-empty ant-empty-normal"
                                      >
                                        <div
                                          class="ant-empty-image"
                                        >
                                          <svg
                                            class="ant-empty-img-simple"
                                            height="41"
                                            viewBox="0 0 64 41"
                                            width="64"
                                            xmlns="http://www.w3.org/2000/svg"
                                          >
                                            <g
                                              fill="none"
                                              fill-rule="evenodd"
                                              transform="translate(0 1)"
                                            >
                                              <ellipse
                                                class="ant-empty-img-simple-ellipse"
                                                cx="32"
                                                cy="33"
                                                rx="32"
                                                ry="7"
                                              />
                                              <g
                                                class="ant-empty-img-simple-g"
                                                fill-rule="nonzero"
                                              >
                                                <path
                                                  d="M55 12.76L44.854 1.258C44.367.474 43.656 0 42.907 0H21.093c-.749 0-1.46.474-1.947 1.257L9 12.761V22h46v-9.24z"
                                                />
                                                <path
                                                  class="ant-empty-img-simple-path"
                                                  d="M41.613 15.931c0-1.605.994-2.93 2.227-2.931H55v18.137C55 33.26 53.68 35 52.05 35h-40.1C10.32 35 9 33.259 9 31.137V13h11.16c1.233 0 2.227 1.323 2.227 2.928v.022c0 1.605 1.005 2.901 2.237 2.901h14.752c1.232 0 2.237-1.308 2.237-2.913v-.007z"
                                                />
                                              </g>
                                            </g>
                                          </svg>
                                        </div>
                                        <p
                                          class="ant-empty-description"
                                        >
                                          No Data
                                        </p>
                                      </div>
                                    </td>
                                  </tr>
                                </tbody>
                              </table>
                            </div>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </main>
        </section>
      </section>
    </div>
  </body>,
  "container": <div>
    <section
      class="ant-layout ant-layout-has-sider"
      data-testid="page-layout"
    >
      <aside
        class="extraExtraLightGrey ant-layout-sider ant-layout-sider-light"
        style="min-height: 100vh; flex: 0 0 200px; max-width: 200px; min-width: 200px; width: 200px;"
      >
        <div
          class="ant-layout-sider-children"
        >
          <div
            class="logo"
          >
            <a
              href="/regulations"
            >
              <svg>
                small_logo.svg
              </svg>
            </a>
          </div>
          <ul
            class="ant-menu sideMenu ant-menu-light ant-menu-root ant-menu-inline"
            role="menu"
          >
            <li
              class="ant-menu-item ant-menu-item-only-child ant-menu-item-selected"
              role="menuitem"
              style="padding-left: 24px;"
            >
              <a
                class="nav-text"
                href="/regulations"
              >
                Overview
              </a>
            </li>
            <li
              class="ant-menu-item ant-menu-item-only-child"
              role="menuitem"
              style="padding-left: 24px;"
            >
              <a
                class="nav-text"
                href="/controls"
              >
                Controls
              </a>
            </li>
            <li
              class="ant-menu-item ant-menu-item-only-child"
              role="menuitem"
              style="padding-left: 24px;"
            >
              <a
                class="nav-text"
                href="/tasks"
              >
                Tasks
              </a>
            </li>
          </ul>
          <div
            class="profile"
            style="width: 175px;"
          >
            <img
              alt="Avatar"
              class="ant-dropdown-trigger"
            />
            <a
              href="/"
            >
              Liis Kindel
            </a>
          </div>
        </div>
      </aside>
      <section
        class="ant-layout white"
        style="background: rgb(255, 255, 255); min-width: 340px;"
      >
        <main
          class="ant-layout-content white"
        >
          <div
            class="site-layout-background"
            style="padding: 24px; min-height: 360px;"
          >
            <div
              class="ant-row ant-row-space-between"
              style="margin: -8px -8px 8px -8px;"
            >
              <div
                class="ant-col ant-col-xs-24 ant-col-sm-16 ant-col-md-10 ant-col-lg-16 ant-col-xl-16 ant-col-xxl-16"
                style="padding: 8px 8px 8px 8px;"
              >
                <h1
                  class="ant-typography"
                >
                  Overview
                </h1>
              </div>
            </div>
            <div
              class="ant-row ant-row-space-between"
              style="margin: -8px -8px 8px -8px;"
            >
              <div
                class="ant-col ant-col-xs-24 ant-col-sm-24 ant-col-md-24 ant-col-lg-24 ant-col-xl-24 ant-col-xxl-24"
                style="padding: 8px 8px 8px 8px;"
              >
                <div
                  class="ant-table-wrapper"
                  style="width: 100%;"
                >
                  <div
                    class="ant-spin-nested-loading"
                  >
                    <div
                      class="ant-spin-container"
                    >
                      <div
                        class="ant-table ant-table-empty"
                      >
                        <div
                          class="ant-table-container"
                        >
                          <div
                            class="ant-table-content"
                          >
                            <table
                              style="table-layout: auto;"
                            >
                              <colgroup />
                              <thead
                                class="ant-table-thead"
                              >
                                <tr>
                                  <th
                                    class="ant-table-cell"
                                  />
                                </tr>
                              </thead>
                              <tbody
                                class="ant-table-tbody"
                              >
                                <tr
                                  class="ant-table-placeholder"
                                >
                                  <td
                                    class="ant-table-cell"
                                  >
                                    <div
                                      class="ant-empty ant-empty-normal"
                                    >
                                      <div
                                        class="ant-empty-image"
                                      >
                                        <svg
                                          class="ant-empty-img-simple"
                                          height="41"
                                          viewBox="0 0 64 41"
                                          width="64"
                                          xmlns="http://www.w3.org/2000/svg"
                                        >
                                          <g
                                            fill="none"
                                            fill-rule="evenodd"
                                            transform="translate(0 1)"
                                          >
                                            <ellipse
                                              class="ant-empty-img-simple-ellipse"
                                              cx="32"
                                              cy="33"
                                              rx="32"
                                              ry="7"
                                            />
                                            <g
                                              class="ant-empty-img-simple-g"
                                              fill-rule="nonzero"
                                            >
                                              <path
                                                d="M55 12.76L44.854 1.258C44.367.474 43.656 0 42.907 0H21.093c-.749 0-1.46.474-1.947 1.257L9 12.761V22h46v-9.24z"
                                              />
                                              <path
                                                class="ant-empty-img-simple-path"
                                                d="M41.613 15.931c0-1.605.994-2.93 2.227-2.931H55v18.137C55 33.26 53.68 35 52.05 35h-40.1C10.32 35 9 33.259 9 31.137V13h11.16c1.233 0 2.227 1.323 2.227 2.928v.022c0 1.605 1.005 2.901 2.237 2.901h14.752c1.232 0 2.237-1.308 2.237-2.913v-.007z"
                                              />
                                            </g>
                                          </g>
                                        </svg>
                                      </div>
                                      <p
                                        class="ant-empty-description"
                                      >
                                        No Data
                                      </p>
                                    </div>
                                  </td>
                                </tr>
                              </tbody>
                            </table>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </main>
      </section>
    </section>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
